Model/DutyCalculator.php

```diff
@@ -116,84 +116,323 @@ public function __construct(
+     * @param Magento\Framework\App\Response $response
+     */
+    public function fillOutQuoteAndSessionOnError($response)
+    {
+        $this->response->setSuccess(false);
+        $this->response->setDuty(0);
+
+        if (isset($response['message'])) { //error message from DHL duty and tax api call?
+            //if yes then do we want to call the API again?
+            //Assuming we do; we are resetting previous country and region value so that we can reenter the blcok to make
+            //DHL API call
+            $this->checkoutSession->setResponseErrorMessage($response['message']);
+            $this->response->setErrorMessage($response['message']);
+            $this->checkoutSession->setPrevCountry('');
+            $this->checkoutSession->setPrevRegion('');
+        }
+    }
+
+    /**
+     * Retrieves duty value for a country or (country, state) combo
+     * and set state of quote, response, session accordingly/as per the value returned
+     * @param $cartId
+     * @param Magento\Quote\Api\Data\AddressInterface $address
+     * @param float $shippingCharge
+     * @param float $duty
+     * @param bool $apply
+     * @param \Magento\Quote\Model\Quote $quote
+     */
+    public function callDHLDutyTaxApi($cartId, $address, $shippingCharge, $duty, $apply, $quote)
+    {
+        $accessToken = $this->getDhlAccessToken();
+
+        //Right moment to make a DHL api call for getting 'Duty and Tax' value
+        if ($accessToken && $accessToken != '') {
+            $this->checkoutSession->setCartId($cartId);
+            //storing selection of country and state to be able to compare it against new set of values
+            //when a potential buyer inputs more data
+            $this->checkoutSession->setPrevCountry($address->getCountryId());
+            $this->checkoutSession->setPrevRegion($address->getRegionCode());
+
+            $request = $this->prepareRequest($shippingCharge, $address);
+            $this->_logger->debug('---------------- Making DHL API call to get Duty and Tax ----------------');
+            $response = $this->getQuote($request, $accessToken);
+
+            $this->checkoutSession->setDhlQuoteId($response['quoteId']);
+
+            //fee value came from DHL
+            if (isset($response['feeTotals'])) {
+                foreach ($response['feeTotals'] as $charge) {
+                    $duty += $charge['value'];
+                }
+                $this->fillOutQuoteAndSessionUsingFeeReturned($duty, $address, $apply, $response);
+
+            } else {
+                $this->fillOutQuoteAndSessionOnError($response);
+
+            }
+        }
+    }
      /**
       * @inheritDoc
       */
     public function getDutyandTax($cartId, $shippingCharge, $shippingMethodCode, $shippingCarrierCode, $address, $apply = false)
     {
+        //this can go somewhere else more appropriate
+        $special_countries = array('CA','BR');
```

Comment:
instead of special_countries which is vague could this array be labeled something like countries_require_state to be more indicative of what this list means?